lib/hammer_cli/defaults.rb

```diff
@@ -0,0 +1,139 @@
+          end
+          rescue NameError => e
+          plugin_prob_message
+          rescue StandardError => e
+          file_not_found_message
+          rescue NotImplementedError => e
+          defaults_not_supproted_by_plugin
+          HammerCLI::EX_CONFIG
+        end
+        HammerCLI::EX_OK
+      end
+    end
+
+    def added_default_message(key, value)
+        print_message(_("Added " + %{key_val} + " default-option with value %{val_val}.") % {:key_val => key.to_s,:val_val => value.to_s } )
+    end
+
+    def plugin_prob_message
+      print_message(_("Couldn't reach the plugin defaults class, eiter the plugin doesn't have defaults class or the plugin misspelled.(example, for Hammer_CLI_Foreman --plugin-name foreman)"))
+    end
+
+    def defaults_not_supproted_by_plugin
+      print_message(_("The param name is not supported by plugin"))
+    end
+
+
+    def bad_input
+      print_message(_("You must specify a value or a plugin name, cant specify both."))
+    end
+
+    def self.file_cant_be_created
+      print_message(_("Couldn't add file to ~/.hammer/cli.modules.d please create the path before defaults will be enabled."))
+    end
+
+    def list_all_defaults_message
+      HammerCLI::Settings.settings[:defaults].each do |key,val|
+          print_message('')
+          print_message(key.to_s+":")
+          print_message(YAML.dump(HammerCLI::Settings.settings[:defaults][key]))
+      end
+    end
+
+    autoload_subcommands
+  end
+
+
+  class Defaults
+    def self.register_provider(provider)
+      Defaults.providers << provider
+    end
+
+    def self.providers
+      @@providers ||= []
+    end
+
+    def self.add_defaults_to_conf(options, provider)
+      create_default_file unless !defaults_file_exists?
+      path = "#{Dir.home}/.hammer/cli.modules.d/defaults.yml"
+      new_file = YAML.load_file(path)
+      options.each do |key, value|
+        key = key.to_sym
+        value.to_s.to_i if value.is_a? Integer
+        new_file[:defaults] = {} if new_file[:defaults].nil?
+        new_file[:defaults][key] = value ? {:value => value, :from_server => false} : {:from_server => true, :provider => provider}
+      end
+      File.open(path,'w') do |h|
+        h.write new_file.to_yaml
+      end
+    end
+
+    #this method will be overriden by plugins who wish to have a defaults params.
+    def self.get_defaults(option)
+      unless HammerCLI::Settings.settings[:defaults].nil? || HammerCLI::Settings.settings[:defaults][option.to_sym].nil?
+        if HammerCLI::Settings.settings[:defaults][option.to_sym][:from_server]
+          value =  HammerCLI::Defaults.providers[HammerCLI::Settings.settings[:defaults][option.to_sym][:provider]].get_defaults(option.to_sym)
+          value
+        else
+          value = HammerCLI::Settings.settings[:defaults][option.to_sym][:value] if defaults_file_exists?
```

Comment:
Looking at the body of `defaults_file_exists?`, ain't it too late to check whether `HammerCLI::Settings.settings[:defaults].nil?` here? You've done it on the line `109`